src/routes/auth/index.ts

```diff
@@ -0,0 +1,147 @@
+import sgMail from '@sendgrid/mail';
+import { type Context, Hono } from 'hono';
+import { ZodError } from 'zod';
+import { generateEmailHtml } from '../../email-templates/confirm-email.ts';
+import { hashPasswordPBKDF2 } from '../../utils/hashPassword.ts';
+import { StatusCodes, type StatusResponse } from '../../utils/responses.ts';
+import { insertProfile } from '../profile/data.ts';
+import { type CreateProfileRequestBody, CreateProfileSchema } from '../profile/validation.ts';
+import { activateProfile, checkEmail, getPassword } from './data.ts';
+import { SignInSchema } from './validate.ts';
+
+export const authRoutes = new Hono();
+
+authRoutes.post('/sign-up', async (context: Context<EnvironmentBindings>) => {
+	const body = await context.req.json();
+	let parsedBody: CreateProfileRequestBody;
+
+	try {
+		parsedBody = CreateProfileSchema.parse(body);
+	} catch (error) {
+		return context.json<StatusResponse>({ message: error?.message ?? 'Invalid input' }, StatusCodes.BAD_REQUEST);
+	}
+
+	const emailExists = await checkEmail(context.env.database, parsedBody.email);
+	if (emailExists) {
+		return context.json({ message: 'Duplicate email' }, StatusCodes.CONFLICT);
+	}
+
+	// Create Profile
+	try {
+		// Generate salt (16 bytes)
+		const salt = crypto.getRandomValues(new Uint8Array(16));
+
+		// Hash password with PBKDF2
+		const hashedPassword = await hashPasswordPBKDF2(parsedBody.password, salt);
+
+		// Convert salt to Base64 for storage
+		const saltBase64 = btoa(String.fromCharCode(...salt));
```

Comment:
This feels red-flaggy. Is `TextEncoder` an option? I wouldn't take my chances with 2 encoding steps that technically should not be compatible. It's possible this is safe but I would rather use a utility that's designed for this specific purpose.

Reply:
`TextEncoder` should be an option on cloudflare workers